### PR TITLE
SOLR-9575 Allow starting with an empty SOLR_HOME

### PR DESCRIPTION
**Pull request by janhoy**
https://issues.apache.org/jira/browse/SOLR-9575

In 8.x this would fail:
```bash
export SOLR_HOME=/tmp/solrhome && mkdir /tmp/solrhome && bin/solr start -c
```
...as you'd  have to pre-initialize your new solr-home with both `solr.xml` and `zoo.cfg`.

This PR will make Solr accept an empty `SOLR_HOME` dir, falling back to the defaults of both files as found in `$SOLR_TIP/server/solr/`. Nothing will be copied into the empty dir, this is a fallback coded into Solr itself. If user wants to customize they can place these files in `SOLR_HOME` and they will be used as before. To test, check out the PR branch, then:

```bash
export SOLR_HOME=/tmp/solrhome
mkdir /tmp/solrhome
bin/solr start -c
grep -E "built-in default|solr.xml|zoo.cfg" server/logs/solr.log 
# [...] SolrXmlConfig Container configuration not found in SOLR_HOME, using built-in default
# [...] SolrXmlConfig Loading container configuration from [...]server/solr/solr.xml
# [...] SolrZkServer Zookeeper configuration not found in /tmp/solrhome, using built-in default
# [...] SolrZkServerProps Reading configuration from: [...]server/solr/zoo.cfg
bin/solr stop
```

Then try out the back-compat override var to trigger 8.x behaviour:

```bash
SOLR_SOLRXML_REQUIRED=true bin/solr start -f 
```
Solr will exit immediately, as before. If Solr is started in cloud mode, Solr will start, since it needs to check for solr.xml in zookeeper, then it will fail early in coreContainer init.

### CODE REVIEW

**Review by dsmiley — COMMENTED** [review]
+1 thanks!
Furthermore, shouldn't init_var_solr no longer copy these two files?

**Comment by janhoy** [comment]
> +1 thanks! Furthermore, shouldn't init_var_solr no longer copy these two files?

Already part of the PR: https://github.com/apache/solr/pull/558/files#diff-7e421bf40f7145b7dcb8933fc47be763b4da343e549b7886d134d56a802bbbbdL59

**Comment by janhoy** [comment]
If no further comments I'll merge this soon.

**Comment by janhoy** [comment]
@HoustonPutman I added you as reviewer as you may have some input from the solr-operator and docker perspective.

In particular, do you see any side effects of this? We should perhaps add an explicit upgrade note for people to *remove* solr.xml and zoo.cfg from their /var/solr/data if they have not customized it on purpose, to let future version upgrades utilize new defaults?

File: solr/bin/solr [hunk 1]
```diff
@@ -2069,8 +2069,11 @@ if [ "$SOLR_MODE" == 'solrcloud' ]; then
     CLOUD_MODE_OPTS+=('-Dbootstrap_confdir=./solr/collection1/conf' '-Dcollection.configName=myconf' '-DnumShards=1')
   fi
 
+  if [ -n "$SOLR_SOLRXML_REQUIRED" ]; then
```
Comment (HoustonPutman) [inline solr/bin/solr:2072]: Shouldn't this and the one be `== true`, like in the `.cmd` version?

**Review by HoustonPutman — COMMENTED** [review]
> I added you as reviewer as you may have some input from the solr-operator and docker perspective.
>
>In particular, do you see any side effects of this? We should perhaps add an explicit upgrade note for people to remove solr.xml and zoo.cfg from their /var/solr/data if they have not customized it on purpose, to let future version upgrades utilize new defaults?

I want to preface this by saying that this is amazing, and a huge step forward. Eventually it will make running Solr in Kubernetes much better. But for now there are complexities.

So the whole `solr.xml` logic in the solr-operator is sadly complex. Well not the logic, but the reasons behind why it's needed. Basically up until 8.8 (so <= 8.7), the Solr Operator cannot correctly manage a Solr Cloud without a custom `solr.xml`. After that there are options from users that require custom entries in the `solr.xml`, many of which are not needed starting in 9.0 luckily.

Once the minimum version that the solr-operator supports is 9.0 (or whatever version this goes into), then we can get rid of a lot of this logic. Until then it will be hard for us to use this. The only way I can think is to add a flag (e.g. `solr9Compatible`) that lets us know that we can use this new functionality, because we don't parse image tags to determine a Solr version.

For users upgrading major versions (when this new logic has been incorporated into the operator), I think I would probably suggest they just create a new cloud and re-index. But I know that's not a suggestion that users appreciate generally. For now you can make a comment on https://github.com/apache/solr-operator/issues/401 or add a note [to this running list](https://github.com/apache/solr-operator/blob/main/dev-docs/changes-for-minimum-solr-version.md), although I've already included the basic information.

File: solr/bin/solr (hunk 1 above)
Comment (janhoy) [reply solr/bin/solr]: Fixed

File: solr/core/src/java/org/apache/solr/core/SolrXmlConfig.java [hunk 2]
```diff
@@ -161,14 +162,21 @@ public static NodeConfig fromConfig(Path solrHome, XmlConfigFile config, boolean
...
-    log.info("Loading container configuration from {}", configFile);
-
     if (!Files.exists(configFile)) {
-      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
-          "solr.xml does not exist in " + configFile.getParent() + " cannot start Solr");
+      if (Boolean.getBoolean("solr.solrxml.required")) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+            "solr.xml does not exist in " + configFile.getParent() + " cannot start Solr");
+      }
+      log.info("Container configuration not found in SOLR_HOME, using built-in default");
```
[the remark this answers is not in the document]
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/core/SolrXmlConfig.java]: Agree, container is something different these days :)

**Comment by janhoy** [comment]
> The only way I can think is to add a flag (e.g. `solr9Compatible`) that lets us know that we can use this new functionality, because we don't parse image tags to determine a Solr version.

Yea, or perhaps a new solr-operator 0.6 could default ot solr9, and instead have a `preSolr9Compatible` config? Or simply start parsing image tag to guess at least the major version? I have a feeling we'll need similar compat breaks for 10.0 as well.

File: solr/solr-ref-guide/src/configuring-solr-xml.adoc [hunk 3]
```diff
@@ -543,7 +553,6 @@ If only `dividend` routing is desired, `hash` may be explicitly set to the empty
 The `<metrics>` element in `solr.xml` allows you to customize the metrics reported by Solr.
 You can define system properties that should not be returned, or define custom suppliers and reporters.
 
-In a default `solr.xml` you will not see any `<metrics>` configuration.
```
Comment (janhoy) [inline solr/solr-ref-guide/src/configuring-solr-xml.adoc:546]: I removed this line from docs, as obviously there is now a `<metrics>` tag in our default solr.xml :)

File: solr/core/src/java/org/apache/solr/cloud/SolrZkServer.java [hunk 4]
```diff
@@ -168,34 +178,29 @@ public void stop() {
...
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
-  private static final Pattern MISSING_MYID_FILE_PATTERN = Pattern.compile(".*myid file is missing$");
 
   String solrPort; // port that Solr is listening on
   String zkRun;
 
   /**
    * Parse a ZooKeeper configuration file
-   * @param path the patch of the configuration file
+   * @param configPath the patch of the configuration file
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/cloud/SolrZkServer.java:187]:
```suggestion
   * @param configPath the path of the configuration file
```

**Pull request merged**